src/compiler/rust.rs

```diff
@@ -0,0 +1,761 @@
+    };
+    let dep_file = temp_dir.path().join("deps.d");
+    let mut cmd = creator.clone().new_command_sync(executable);
+    cmd.args(&arguments)
+        .args(&["--emit", "dep-info"])
+        .arg("-o")
+        .arg(&dep_file)
+        .current_dir(cwd);
+    trace!("[{}]: get dep-info: {:?}", crate_name, cmd);
+    let dep_info = run_input_output(cmd, None);
+    // Parse the dep-info file, then hash the contents of those files.
+    let pool = pool.clone();
+    let cwd = cwd.to_owned();
+    let crate_name = crate_name.to_owned();
+    Box::new(dep_info.and_then(move |output| -> SFuture<_> {
+        if output.status.success() {
+            match parse_dep_file(&dep_file, &cwd) {
+                Ok(files) => {
+                    trace!("[{}]: got {} source files from dep-info in {}", crate_name,
+                           files.len(), fmt_duration_as_secs(&start.elapsed()));
+                    // Just to make sure we capture temp_dir.
+                    drop(temp_dir);
+                    hash_all(files, &pool)
+                }
+                Err(e) => return Box::new(future::err(e)),
+            }
+        } else {
+            Box::new(future::err("Failed run rustc --dep-info".into()))
+        }
+    }))
+}
+
+/// Parse dependency info from `file` and return a Vec of files mentioned.
+/// Treat paths as relative to `cwd`.
+fn parse_dep_file<T, U>(file: T, cwd: U) -> Result<Vec<String>>
+    where T: AsRef<Path>,
+          U: AsRef<Path>,
+{
+    let mut f = File::open(file)?;
+    let mut deps = String::new();
+    f.read_to_string(&mut deps)?;
+    Ok(parse_dep_info(&deps, cwd))
+}
+
+fn parse_dep_info<T>(dep_info: &str, cwd: T) -> Vec<String>
+    where T: AsRef<Path>
+{
+    let cwd = cwd.as_ref();
+    //TODO: sort files.
+    let mut deps = dep_info.lines()
+        // The first line is the dependencies on the dep file itself.
+        .skip(1)
+        .filter_map(|l| if l.is_empty() { None } else { l.split(":").next() })
+        .map(|s| cwd.join(s).to_string_lossy().into_owned())
+        .collect::<Vec<_>>();
+    deps.sort();
+    deps
+}
+
+/// Run `rustc --print file-names` to get the outputs of compilation.
+fn get_compiler_outputs<T>(creator: &T, executable: &str, arguments: &[String], cwd: &str) -> SFuture<Vec<String>>
+    where T: CommandCreatorSync,
+{
+    let mut cmd = creator.clone().new_command_sync(executable);
+    cmd.args(&arguments)
+        .args(&["--print", "file-names"])
+        .current_dir(cwd);
+    if log_enabled!(Trace) {
+        trace!("get_compiler_outputs: {:?}", cmd);
+    }
+    let outputs = run_input_output(cmd, None);
+    Box::new(outputs.and_then(move |output| -> Result<_> {
+        if output.status.success() {
+            let outstr = String::from_utf8(output.stdout).chain_err(|| "Error parsing rustc output")?;
+            Ok(outstr.lines().map(|l| l.to_owned()).collect())
+        } else {
+            bail!("Failed to run `rustc --print file-names`");
```

Comment:
I wonder if this may actually make a good helper as part of `run_input_output`, having a method to run a command and auto-check `output.status.success()` and bundle up all the relevant error info in the case of failure.